crates/holochain_deterministic_integrity/src/link.rs

```diff
@@ -1,69 +1,92 @@
-use core::array::IntoIter;
-use core::ops::RangeBounds;
+use holochain_integrity_types::LinkTypeFilter;
+use holochain_wasmer_guest::WasmError;
 
 use crate::prelude::*;
 
-pub use hdk_derive::hdk_link_types;
-
 #[cfg(doc)]
 pub mod examples;
 
-/// A helper trait for creating [`LinkTypeRanges`] that match the local zome's
-/// type scope.
-///
-/// This is implemented by the [`hdk_link_types`] proc_macro.
-pub trait LinkTypesHelper<const LEN: usize>: EnumLen
+pub trait LinkTypeFilterExt {
+    fn try_into_filter(self) -> Result<LinkTypeFilter, WasmError>;
+}
+
+impl LinkTypeFilterExt for core::ops::RangeFull {
+    fn try_into_filter(self) -> Result<LinkTypeFilter, WasmError> {
+        let out = zome_info()?.zome_types.links.dependencies().collect();
+        Ok(LinkTypeFilter::Dependencies(out))
+    }
+}
+
+impl LinkTypeFilterExt for LinkTypeFilter {
+    fn try_into_filter(self) -> Result<LinkTypeFilter, WasmError> {
+        Ok(self)
+    }
+}
+
+impl<T, E> LinkTypeFilterExt for Vec<T>
 where
-    Self: Into<LocalZomeTypeId>,
-    Self: std::fmt::Debug + Clone + Copy + Sized + PartialEq + PartialOrd + 'static,
+    T: TryInto<ScopedLinkType, Error = E>,
+    WasmError: From<E>,
 {
-    /// Create a [`LinkTypeRanges`] from a range of this traits implementor.
-    fn range(
-        range: impl RangeBounds<Self> + 'static + std::fmt::Debug,
-    ) -> Box<dyn FnOnce() -> Result<LinkTypeRanges, WasmError>> {
-        let zome_types = zome_info().map(|t| t.zome_types);
-        let f = move || {
-            let zome_types = zome_types?;
+    fn try_into_filter(self) -> Result<LinkTypeFilter, WasmError> {
+        let mut vec = self
+            .into_iter()
+            .map(TryInto::<ScopedLinkType>::try_into)
+            .collect::<Result<Vec<_>, _>>()?;
+        vec.sort_unstable();
+        let vec = vec.into_iter().fold(
+            Vec::new(),
+            |mut out: Vec<(ZomeId, Vec<LinkType>)>,
+             ScopedZomeType {
+                 zome_id,
+                 zome_type: link_type,
+             }| {
+                match out.last_mut() {
+                    Some(l) if l.0 == zome_id => l.1.push(link_type),
+                    _ => out.push((zome_id, vec![link_type])),
+                }
+                out
+            },
+        );
```

Comment:
maybe some code docs in here to explain why we're sorting / folding / etc?

Reply:
Will do